samples/bluetooth/peripheral_hids_keyboard/src/main.c

```diff
@@ -0,0 +1,629 @@
+					SCAN_CODE_POS)
+
+#define CAPS_LOCK_LED        LED0_GPIO_PIN
+#define SHIFT_BUTTON         SW1_GPIO_PIN
+#define SHIFT_BUTTON_PORT    1
+#define SHIFT_BUTTON_PRESSED 0
+
+/* HIDs queue size. */
+#define HIDS_QUEUE_SIZE 10
+
+/* HIDS instance. */
+HIDS_DEF(hids_obj,
+	 OUTPUT_REPORT_MAX_LEN,
+	 INPUT_REPORT_KEYS_MAX_LEN);
+
+/* Required security level */
+#ifdef CONFIG_BT_GATT_HIDS_SECURITY_LEVEL_LOW
+static const bt_security_t sec_level = BT_SECURITY_LOW;
+#elif CONFIG_BT_GATT_HIDS_SECURITY_LEVEL_MED
+static const bt_security_t sec_level = BT_SECURITY_MEDIUM;
+#elif CONFIG_BT_GATT_HIDS_SECURITY_LEVEL_HIGH
+static const bt_security_t sec_level = BT_SECURITY_HIGH;
+#elif CONFIG_BT_GATT_HIDS_SECURITY_LEVEL_FIPS
+static const bt_security_t sec_level = BT_SECURITY_FIPS;
+#else
+static const bt_security_t sec_level = BT_SECURITY_LOW;
+#endif
+
+struct key_pack {
+	void *private;
+	u8_t *elements;
+	u8_t len;
+};
+
+/* Keyboard queue. */
+K_MSGQ_DEFINE(hids_queue,
+	      sizeof(struct key_pack),
+	      HIDS_QUEUE_SIZE,
+	      4);
+
+static const struct bt_data ad[] = {
+	BT_DATA_BYTES(BT_DATA_GAP_APPEARANCE,
+		      (CONFIG_BT_DEVICE_APPEARANCE >> 0) & 0xff,
+		      (CONFIG_BT_DEVICE_APPEARANCE >> 8) & 0xff),
+	BT_DATA_BYTES(BT_DATA_FLAGS, (BT_LE_AD_GENERAL | BT_LE_AD_NO_BREDR)),
+	BT_DATA_BYTES(BT_DATA_UUID16_ALL,
+		      0x12, 0x18,       /* HID Service */
+		      0x0f, 0x18),      /* Battery Service */
+
+};
+
+static const struct bt_data sd[] = {
+	BT_DATA(BT_DATA_NAME_COMPLETE, DEVICE_NAME, DEVICE_NAME_LEN),
+};
+
+static u8_t hello_world_str[] = {
+	0x0b,	/* Key h */
+	0x08,	/* Key e */
+	0x0f,	/* Key l */
+	0x0f,	/* Key l */
+	0x12,	/* Key o */
+	0x28,	/* Key Return */
+};
+
+static u8_t caps_on_key_scan_str[] = {
+	0x06,	/* Key C */
+	0x04,	/* Key a */
+	0x13,	/* Key p */
+	0x16,	/* Key s */
+	0x12,	/* Key o */
+	0x11,	/* Key n */
+};
+
+static u8_t caps_off_key_scan_str[] = {
+	0x06,	/* Key C */
+	0x04,	/* Key a */
+	0x13,	/* Key p */
+	0x16,	/* Key s */
+	0x12,	/* Key o */
+	0x09,	/* Key f */
+};
+
+static struct device		*led_dev;
+static struct device		*gpio_devs[4];
+static struct gpio_callback	gpio_cbs[4];
+static bool			caps_on;
+
+static struct conn_mode {
+	struct bt_conn *conn;
+	bool in_boot_mode;
+} conn_mode[CONFIG_BT_GATT_HIDS_MAX_CLIENT_COUNT];
+
+
+static void advertising_start(void)
+{
+	int err = bt_le_adv_start(BT_LE_ADV_CONN, ad, ARRAY_SIZE(ad),
+				  sd, ARRAY_SIZE(sd));
+
+	if (err) {
+		printk("Advertising failed to start (err %d)\n", err);
+		return;
+	}
+
+	printk("Advertising successfully started\n");
+}
+
+
+static void connected(struct bt_conn *conn, u8_t err)
+{
+	char addr[BT_ADDR_LE_STR_LEN];
+
+	bt_addr_le_to_str(bt_conn_get_dst(conn), addr, sizeof(addr));
+
+	if (err) {
+		printk("Failed to connect to %s (%u)\n", addr, err);
+		return;
+	}
+
+	printk("Connected %s\n", addr);
+
+	if (IS_ENABLED(CONFIG_BT_GATT_HIDS_SECURITY_ENABLED)) {
```

Comment:
There is a PR open to let the central side initiate security instead: https://github.com/NordicPlayground/fw-nrfconnect-nrf/pull/350.

Reply:
Ok... I will update it. Also I have created a task to allow permission changes inside HIDS reports.